specification.md

```diff
@@ -0,0 +1,282 @@
+
+```bash
+make all # equal to make build
+make build
+make build-local
+make build-in-container
+```
+
+Rules for Makefile targets:
+
+-   `make all` MUST be equal to `make build`
+-   `make build` means to compile you project. The Makefile MUST support the ability to compile the project on localhost and in docker container.
+-   For convenience, the Makefile SHOULD support the ability to build targets directories directly. For example:
+
+```bash
+make cmd/controller # equal to make build WHAT=cmd/controller
+```
+
+#### Options
+
+| Name        | Usage                                    | Type   | Default | Severity |
+| ----------- | ---------------------------------------- | ------ | ------- | -------- |
+| LOCAL_BUILD | If set true, project will be built on local machine, otherwise, built in docker. | string | true    | REQUIRED |
+| WHAT        | Directory names to build.  If any of these directories has a main package, the build will produce executable files under bin/. e.g `cmd/controller` | string |         | REQUIRED |
+
+#### Example
+
+The most basic rules:
+
+```bash
+# make all targets
+make all
+make build
+make build LOCAL_BUILD=true
+
+# make cmd/controller on localhost
+make build WHAT=cmd/controller LOCAL-BUILD=true
+# make cmd/controller in docker
+make build WHAT=cmd/controller LOCAL-BUILD=false
+```
+
+the followings are extensions for the basic way
+
+```bash
+# make all targets on localhost
+make build-local
+# make all targets in docker contaienr
+make build-in-container
+# make cmd/controller
+make cmd/controller
+
+# more ...
+make build-local WHAT=cmd/controller
+make build-in-container WHAT=cmd/controller
+make cmd/controller LOCAL_BUILD=false
+```
+
+#### Compile different language
+
+For more information about different language compile specification:
+
+-   [Golang Compile](#golangCompile)
+
+
+### Unit Test
+
+The Makefile MUST support the following targets for unit test
+
+```bash
+make unittest
+```
+
+#### Test different language 
+
+For more information about different language unittest specification:
+
+-   [Golang Unittest](#golangUnittest)
+
+### Docker
+
+The Makefile MUST support the following targets for docker build.
+
+```bash
+make container
+make push
+```
+
+Rules for Makefile targets:
+
+-   The Makefile MUST support the ability to build or push docker images for multiple registries, and modify docker image by adding prefix and suffix.
+-   For convenience, the Makefile SHOULD support the ability to build targets directories which contain Dockerfiles directly. For example:
+
+```bash
+make build/controller # equal to make container WHAT=build/controller
+```
+
+#### Options
+
+| Name                 | Usage                                    | Type   | Default | Severity |
+| -------------------- | ---------------------------------------- | ------ | ------- | -------- |
+| DOCKER_REGISTRIES    | Docker registries to push                | array  |         | REQUIRED |
```

Comment:
Can you give some examples on how to use the options?

Reply:
OK